data-science-onramp/data-ingestion/setup.py

```diff
@@ -0,0 +1,163 @@
+
+
+def user_type(user):
+    '''Manipulates the user type string'''
+    return random.choice([user, user.upper(), user.lower(),
+                          "sub" if user == "Subscriber" else user,
+                          "cust" if user == "Customer" else user])
+
+
+def gender(s):
+    '''Manipulates the gender string'''
+    return random.choice([s.upper(), s.lower(),
+                         s[0].upper() if len(s) > 0 else "",
+                         s[0].lower() if len(s) > 0 else ""])
+
+
+def convert_angle(angle):
+    '''Converts long and lat to DMS notation'''
+    degrees = int(angle)
+    minutes = int((angle - degrees) * 60)
+    seconds = int((angle - degrees - minutes/60) * 3600)
+    new_angle = str(degrees) + u"\u00B0" + \
+        str(minutes) + "'" + str(seconds) + '"'
+    return random_select([str(angle), new_angle], [0.55, 0.45])
+
+
+# This function is nested since a UserDefinedFunction is
+# expected to take a single argument
+def dirty_data(proc_func, allow_none):
+    '''Master function returns a user defined function
+    that transforms the column data'''
+    def udf(col_value):
+        random.seed(hash(col_value) + time_ns())
+        if col_value is None:
+            return col_value
+        elif allow_none:
+            return random_select([None, proc_func(col_value)],
+                                 [0.05, 0.95])
+        else:
+            return proc_func(col_value)
+    return udf
+
+
+# This function is required because we need to apply a
+# function for every column and some columns do not change
+def identity(x):
+    return x
+
+
+def write_to_bigquery(spark, df):
+    '''Write a dataframe to BigQuery'''
+
+    # Create BigQuery Dataset
+    client = bigquery.Client()
+    dataset_id = f'{client.project}.new_york_citibike_trips'
+    dataset = bigquery.Dataset(dataset_id)
+    dataset.location = "US"
+    dataset = client.create_dataset(dataset)
+
+    # Saving the data to BigQuery
+    spark.conf.set('temporaryGcsBucket', BUCKET_NAME)
+
+    df.write.format('bigquery') \
+        .option('table', dataset_id + ".RAW_DATA") \
+        .save()
+
+
+def main():
+    # Create a SparkSession under the name "setup". Viewable via the Spark UI
+    spark = SparkSession.builder.appName("setup").getOrCreate()
+
+    upload = True  # Whether to upload data to BigQuery
+
+    # Check whether or not results should be uploaded
+    if len(sys.argv) > 1:
+        upload = False
+        print("Not uploading results to BigQuery")
+    else:
+        print("Results will be uploaded to BigQuery")
+
+    # Check if table exists
+    try:
+        df = spark.read.format('bigquery').option('table', TABLE).load()
+    except Py4JJavaError:
+        print(f"{TABLE} does not exist. ")
+        return
+
+    # Declare data transformations for each column in dataframe
+    udfs = [
+        (dirty_data(trip_duration, True), StringType()),  # tripduration
+        (dirty_data(identity, True), StringType()),  # starttime
```

Comment:
Unless the `identity` function is used elsewhere within the job, you can replace it by using a lambda function `_ = lambda x: x` that you declare just before you create `udfs`. You can then replace every instance of `identity` with `_`.